doc/developer/overview.md

```diff
@@ -0,0 +1,116 @@
+
+# Materialize architecture
+
+Materialize presents as a SQL-ish system that allows users to name, manipulate, query, and produce continually changing data.
+To perform these tasks Materialize is broken into several components that interact through relatively narrow interfaces.
+These interfaces both provide and require guarantees, which are important for Materialize to continue to work correctly and to minimize the cognitive burden of working *within* Materialize.
+
+```
+SQL front-end   <-\   Coordinator       Dataflow
+SQL front-end   <-+->  *Optimizer <-->   *Sources
+SQL front-end   <-/    *Catalog          *Sinks
+```
+
+This document means to frame these components, the properties they require, and the guarantees they provide.
+
+## SQL front end
+
+<TBD: @benesch>
+
+### Lowering
+
+## Coordinator
+
+The coordinator is the "brains" of Materialize.
+It records the "metadata" of what data users can access, and the properties those data have.
+It manages the execution of queries, providing instruction to the dataflow layer of what computations to perform.
+It selects timestamps at which queries will execute, which is a critical component of providing the experience of consistency (here: most likely "linearizability").
+
+### Catalog manipulation
+
+<TBD: @benesch>
+
+### Timestamp determination
```

Comment:
I've marked this as `TBD @mjibson`. I think we do care but with e.g. `VOLATILE` sources we have relaxed properties.

Reply:
We should add serializability, so I'd change the line to:

```
In particular we aim to provide the properties of linearizability and serializability.
```

I personally would link to jepsen instead of wikipedia, but probably doesn't matter. http://jepsen.io/consistency/models/serializable and http://jepsen.io/consistency/models/linearizable.

To address volatility we should add:

```
These guarantees are dependent on the source's [volatility](https://materialize.com/docs/overview/volatility/), and if data changes between restarts, so could our query results.
```